Add Mailing To User After Account Creation

GuenoleCHERUBIN opened the pull request:
Added the function to create a string of X
random characters, the X value has to be pasted
on funciton calling. This function is used
to to create a unique token for a user.
Added phpmailer, to send an email.
Phpmailer is bound to mailhog for now.
Added the token column in the database.
Added a new exception that is thrown if
we can't create a token.

GuenoleCHERUBIN committed:
Add Mailing To User After Account Creation

Added the function to create a string of X
random characters, the X value has to be pasted
on funciton calling. This function is used
to to create a unique token for a user.
Added phpmailer, to send an email.
Phpmailer is bound to mailhog for now.
Added the token column in the database.
Added a new exception that is thrown if
we can't create a token.

EmileCalixte commented on controllers/ManagementController.php:
```diff
@@ -172,7 +175,7 @@ public function actionCreateAccount()
 
         if($form->load($_POST) && $form->validate()) {
-            $this->createAccount($form);
+                $this->createAccount($form);
```
🤔

GuenoleCHERUBIN replied:
To be completely honest with you, i don't where this ones coming from.

EmileCalixte commented on controllers/ManagementController.php:
```diff
@@ -195,11 +201,24 @@ private function createAccount(CreateAccountForm $form)
         $user->is_validated = false;
         $user->is_active = true;
         $user->role = $form->role;
+        $token_try = 0;
```
This variable name is not camelCase

alexisljn replied:
I agree

EmileCalixte commented on controllers/ManagementController.php:
```diff
@@ -195,11 +201,24 @@ private function createAccount(CreateAccountForm $form)
+
+        if($token_try == 10) {
+            throw new CannotCreateTokenException();
+        }
```
Even if it's unlikely, there may be a case where you throw this exception when the token creation has succeeded :
- Your loop failed 9 times, so $tokenTry is 9
- You make a 10th iteration, you succeed to create a user token
- You increments $tokenTry (line 207)
- After the loop, $tokenTry is 10, so you throws the exception, whereas the token is well created

GuenoleCHERUBIN replied:
That's true, I'll try to figure out something more accurate, and that works in all cases

EmileCalixte commented on controllers/ManagementController.php:
```diff
@@ -283,5 +302,37 @@ private function updateSocialMedia(ModifySocialMediaInformationsForm $formModify
+        $mail->isHTML(false);
+        $mail->CharSet = 'UTF-8';
+
+        $mail->Subject = "Account's creation for " . $user->firstname . " " . $user->lastname;
```
I suggest "Activate your Collab'media accoun" as mail subject

GuenoleCHERUBIN replied:
Yeah seems more accurate

EmileCalixte commented on controllers/ManagementController.php:
```diff
@@ -283,5 +302,37 @@ private function updateSocialMedia(ModifySocialMediaInformationsForm $formModify
+                return false;
+            }
+        } catch(Exception $e) {
+            return false;
```
It is a bad practice to return false when an exception is caught, because you lose the error information. Your mailToUserPasswordCreation method should simply throw the exception, and you should optionally catch it in the main function which is calling this method.

GuenoleCHERUBIN committed:
Changes After Comments

Added a new exception after comment.
Changed the name of a variable so it's in camelCAse

GuenoleCHERUBIN committed:
Changes After Comment

alexisljn commented on controllers/ManagementController.php:
```diff
@@ -283,5 +303,34 @@ private function updateSocialMedia(ModifySocialMediaInformationsForm $formModify
+        $mail->CharSet = 'UTF-8';
+
+        $mail->Subject = "Activate your Collab'media account";
+        $mail->Body = 'Click on the following link to create your password : http://127.0.0.1:8000/site/validate-account/' . $user->token ;
```
actionValidateAccount() will be created later, that's right ?

alexisljn reviewed (commented):
actionValidateAccount() will be created later, that's right ?

EmileCalixte commented on controllers/ManagementController.php:
```diff
@@ -201,14 +202,14 @@ private function createAccount(CreateAccountForm $form)
+        } while (!is_null(User::findOne(['token' => $token])) AND $tokenTry < 10);
 
-        if($token_try == 10) {
+        if($tokenTry == 10 OR !is_null(User::findOne(['token' => $token]))) {
```
That's not goot, it should be AND, not OR. I suggest you move this condition inside the do-while (just before calling `$this->createUserToken()`), and remove the `AND $tokenTry < 10` from your while condition.

GuenoleCHERUBIN committed:
Change The Way The Token Creation Is Handled

Changed the way we handle the token creation

alexisljn reviewed (approved):
nice dude

EmileCalixte reviewed (approved):
Perfect !
